File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<deque>
#include<iostream>

using namespace std;

pair<int, int> direction[4] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
vector<vector<int>> visited;

int solution(vector<vector<int>> maps)
{
    for (int i = 0; i < maps.size(); i++){
        vector<int> row;
        for (int j = 0; j < maps[0].size(); j++){
            row.push_back(0);
        }
        visited.push_back(row);
    }
    
    deque<pair<int, int>> dq;
    dq.push_back({0, 0});
    visited[0][0] = 1;
    while (!dq.empty()){
        pair<int, int> curr = dq.front();
        dq.pop_front();
        for (pair<int, int> d: direction){
            pair<int, int> next = {curr.first + d.first, curr.second + d.second};
            if (
                next.first >= 0 && next.first < maps.size()
                && next.second >= 0 && next.second < maps[0].size()
                && maps[next.first][next.second] == 1
                && visited[next.first][next.second] == 0
            ){
                dq.push_back(next);
                visited[next.first][next.second] = visited[curr.first][curr.second] + 1;
            }
        }
    }
    if (visited[maps.size()-1][maps[0].size()-1] == 0){
        return -1;
    }
    return visited[maps.size()-1][maps[0].size()-1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;

        int[][] visited = new int[n][m];

        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{0, 0});
        visited[0][0] = 1;

        while (!q.isEmpty()) {
            int[] curr = q.poll();

            for (int[] dir : dirs) {
                int nx = curr[0] + dir[0];
                int ny = curr[1] + dir[1];

                // map 밖인지 확인
                if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;

                // 벽인지 확인
                if (maps[nx][ny] == 0) continue;

                // 이동
                if (maps[nx][ny] == 1 && visited[nx][ny] == 0) {
                    visited[nx][ny] = visited[curr[0]][curr[1]] + 1;
                    q.add(new int[]{nx, ny});
                }

            }
        }


        if (visited[n - 1][m - 1] == 0) {
            return -1;
        } else {
            return visited[n - 1][m - 1];
        }
    }
}